### PR TITLE
[WIP] Track WAL in MANIFEST

### PR DESCRIPTION
**Pull request by cheng-chang**
This PR adds WAL lifecycle related information to VersionEdit.
When a WAL is closed because the corresponding memtable is flushed, the WAL's log number and size are tracked in VersionEdit. When a WAL is archived or deleted, the event is also tracked in VersionEdit.

An alternative design is to track WAL's file path, instead of the log number. This PR prefers to track log number to save disk space occupied by MANIFEST.

Another design choice is to just track added WAL and deleted WAL, for archived WAL, treat is as deleting the old WAL and adding a new WAL in the archive directory. This PR prefers to introduce a specific entry for archived WAL, because it only needs to track the log number in one entry instead of two entries (delete and add), which saves disk space of MANIFEST.

Follow up PRs will:
1. apply the WAL related VersionEdits to MANIFEST.
2. create WAL related VersionEdits in relevant code paths of WAL's lifecycle.
3. compute and track checksum of the WALs.

Test Plan:
A set of new unit tests are added to version_edit_test.cc.

### CODE REVIEW

**Comment by ltamasi** [comment]
Just wanted to clarify what our goal is here... If we want to make sure that WAL files don't go missing from the filesystem, I think we should register them in the MANIFEST when they are first opened (i.e. upon DB open or CF flush) as opposed to when they are closed. (Of course, no size/checksum information is available at that time, so we would need to write a second record with that information upon close.) Also note we already have some information in the MANIFEST about which WAL files can be deleted (see `log_number_` and friends in `VersionEdit`).

File: db/version_edit.h [hunk 1]
```diff
@@ -241,6 +241,57 @@ struct LevelFilesBrief {
   }
 };
 
+using WalNumber = uint64_t;
+
+// Information of a new WAL tracked by VersionEdit.
+class AddedWal {
+ public:
+  AddedWal() = default;
```
Comment (ltamasi) [inline db/version_edit.h]: Note that this leaves `number_` and `bytes_` essentially uninitialized; consider using in-class initializers. (Affects `DeletedWal` too.)

File: db/version_edit.h [hunk 2]
```diff
@@ -241,6 +241,57 @@ struct LevelFilesBrief {
...
+  Status DecodeFrom(Slice* src);
+
+  std::string DebugString() const;
+
+ private:
+  WalNumber number_;
+  uint64_t bytes_;
+};
+
+JSONWriter& operator<<(JSONWriter& jw, const AddedWal& wal);
```
Comment (ltamasi) [inline db/version_edit.h]: We should probably declare the other `operator<<` too here (also affects the other class).

File: db/version_edit.h [hunk 3]
```diff
@@ -413,6 +464,26 @@ class VersionEdit {
   bool IsInAtomicGroup() const { return is_in_atomic_group_; }
   uint32_t GetRemainingEntries() const { return remaining_entries_; }
 
+  // Add a completed WAL.
+  void AddWal(const AddedWal& wal) { new_wals_.push_back(wal); }
+
+  // Get a list of non-archived WALs in the order of AddWal.
+  std::vector<AddedWal> GetNewWals() { return new_wals_; }
```
Comment (ltamasi) [inline db/version_edit.h]: This takes a copy; return a const-ref instead? Also, the method itself could be marked `const`. (Affects the other two getters too.)

File: db/version_edit_test.cc [hunk 4]
```diff
@@ -309,6 +309,109 @@ TEST_F(VersionEditTest, BlobFileAdditionAndGarbage) {
...
+}
+
+TEST_F(VersionEditTest, AddedWal) {
+  constexpr uint64_t kLogNumber = 10;
+  constexpr uint64_t kBytes = 100;
+
+  VersionEdit edit;
+  edit.AddWal(AddedWal(kLogNumber, kBytes));
+
+  auto wals = edit.GetNewWals();
```
Comment (ltamasi) [inline db/version_edit_test.cc:339]: `const auto& wals` ? Also, `const AddedWal& wal` below. (Affects the other two test cases as well.)

File: db/version_edit.cc [hunk 5]
```diff
@@ -302,6 +378,22 @@ bool VersionEdit::EncodeTo(std::string* dst) const {
     PutVarint32(dst, kInAtomicGroup);
     PutVarint32(dst, remaining_entries_);
   }
+
+  for (const AddedWal& wal : new_wals_) {
+    PutVarint32(dst, kNewWal);
+    wal.EncodeTo(dst);
+  }
+
+  for (WalNumber log_number : archived_wals_) {
```
Comment (ltamasi) [inline db/version_edit.cc:351]: Assuming we want to track these events, we could consider introducing a dedicated class `ArchivedWal` that would encapsulate the related logic.

File: db/version_edit.h [hunk 6]
```diff
@@ -241,6 +241,57 @@ struct LevelFilesBrief {
...
+};
+
+JSONWriter& operator<<(JSONWriter& jw, const AddedWal& wal);
+
+// Information of a deleted WAL.
+class DeletedWal {
+ public:
+  DeletedWal() = default;
+
+  DeletedWal(WalNumber number, bool archived)
```
Comment (ltamasi) [inline db/version_edit.h:276]: Again, this is contingent on what we want to track but `archived` could be inferred from the presence of `kArchivedWal`, right?

**Comment by cheng-chang** [comment]
Closing the PR for now since it's still WIP.

**Pull request closed without merging**